### PR TITLE
[editor] Get user stats from server

### PR DESCRIPTION
**Pull request by mgsergio**
do not merge потму, что надо потестировать с gui.

@ygorshenin @yunikkk @igrechuhin

### CODE REVIEW

File: editor/user_stats.hpp [hunk 1]
```diff
@@ -1,30 +1,66 @@
...
   int32_t m_rank;
+  time_t m_updateTime;
+};
+
+class UserStatsLoader
+{
+public:
+  UserStatsLoader();
+
+  using TOnUpdateColback = function<void()>;
```
Comment (igrechuhin) [inline editor/user_stats.hpp]: callback

File: editor/editor_tests/user_stats_test.cpp [hunk 2]
```diff
@@ -2,19 +2,49 @@
...
+    TEST(userStats->GetChangesCount(changesCount), ());
+
+    TEST_GREATER_OR_EQUAL(rank, 5800, ());
+    TEST_EQUAL(changesCount, 2, ());
+  }
+
+  {
+    UserStatsLoader statsLoader;
+
+    TEST_EQUAL(statsLoader.GetUserName(), kUserName, ());
```
Comment (ygorshenin) [inline editor/editor_tests/user_stats_test.cpp]: Why do we need 2 tests? Second tests looks pretty similar to the first one.

File: editor/user_stats.cpp [hunk 3]
```diff
@@ -1,41 +1,70 @@
...
 using TRequest = alohalytics::HTTPClientPlatformWrapper;
 
-
 namespace
 {
 string const kUserStatsUrl = "http://py.osmz.ru/mmwatch/user?format=xml";
-auto constexpr kUninitialized = -1;
+int32_t constexpr kUninitialized = -1;
+
+auto constexpr kSettingsUserName = "LastLoginedUser";
```
Comment (ygorshenin) [inline editor/user_stats.cpp]: Logined? Maybe Logged?

Comment (ygorshenin) [inline editor/user_stats.cpp]: Or just LastUser?

File: editor/user_stats.cpp [hunk 4]
```diff
@@ -1,41 +1,70 @@
...
-bool UserStats::IsRankInitialized() const
+// UserStatsLoader ---------------------------------------------------------------------------------
+
+UserStatsLoader::UserStatsLoader()
 {
-  return m_rank != kUninitialized;
+  if (!LoadFromSettings())
+    LOG(LINFO, ("There is no cached user stats in settings"));
+  else
+    LOG(LINFO, ("User stats was loaded successfully"));
```
Comment (ygorshenin) [inline editor/user_stats.cpp]: were

File: editor/user_stats.cpp [hunk 5]
```diff
@@ -1,41 +1,70 @@
...
 }
 
-bool UserStats::IsRankInitialized() const
+// UserStatsLoader ---------------------------------------------------------------------------------
+
+UserStatsLoader::UserStatsLoader()
 {
-  return m_rank != kUninitialized;
+  if (!LoadFromSettings())
+    LOG(LINFO, ("There is no cached user stats in settings"));
```
Comment (ygorshenin) [inline editor/user_stats.cpp]: are

File: editor/user_stats.cpp [hunk 6]
```diff
@@ -1,41 +1,70 @@
...
 }
 
-bool UserStats::Update()
+bool UserStatsLoader::Update(string const & userName)
 {
+  {
+    lock_guard<mutex> g(m_mutex);
+    m_userName = userName;
+  }
+
```
Comment (ygorshenin) [inline editor/user_stats.cpp]: Data race below on m_userName.

Comment (ygorshenin) [inline editor/user_stats.cpp]: BTW, there are several places where m_userName is used without lock.

File: editor/user_stats.cpp [hunk 7]
```diff
@@ -60,9 +89,100 @@ bool UserStats::Update()
...
+void UserStatsLoader::Update(string const & userName, TOnUpdateColback const & fn)
+{
+  {
+    lock_guard<mutex> g(m_mutex);
+    auto const nothingToUpdate = m_userStats && m_userName == userName && m_userStats &&
+                                 difftime(m_lastUpdate, time(nullptr)) < kOneHour;
+
+    if (nothingToUpdate)
+    {
+      GetPlatform().RunOnGuiThread(fn);
```
Comment (ygorshenin) [inline editor/user_stats.cpp]: Are you sure that you want to call RunOnGuiThread() under protection of m_mutex? Why not to do that without it?

File: editor/user_stats.cpp [hunk 8]
```diff
@@ -60,9 +89,100 @@ bool UserStats::Update()
...
+      return;
+    }
+  }
+
+  thread([this, userName, &fn] {
+    if (Update(userName))
+      GetPlatform().RunOnGuiThread(fn);
+  }).detach();
```
Comment (ygorshenin) [inline editor/user_stats.cpp]: std::async() ?

File: map/framework.hpp [hunk 9]
```diff
@@ -675,15 +675,24 @@ class Framework
...
-  bool UpdateUserStats(string const & userName);
-  void DropUserStats() { m_userStats = nullptr; }
+  // User statistics.
+
+  shared_ptr<editor::UserStats const> GetUserStats(string const & userName) const
+  {
+    return m_userStatsLoader.GetStats(userName);
+  }
+
+  // Reads user stats from server or gets it from cache calls reader on success.
```
Comment (ygorshenin) [inline map/framework.hpp]: or gets it from cache and calls |fn| on success.

Comment (ygorshenin) [inline map/framework.hpp]: BTW, why not to call fn in case of failure?

File: editor/user_stats.hpp (hunk 1 above)
Comment (mgsergio) [inline editor/user_stats.hpp]: Done.

File: editor/user_stats.cpp [hunk 10]
```diff
@@ -1,41 +1,70 @@
...
-
 namespace
 {
 string const kUserStatsUrl = "http://py.osmz.ru/mmwatch/user?format=xml";
-auto constexpr kUninitialized = -1;
+int32_t constexpr kUninitialized = -1;
+
+auto constexpr kSettingsUserName = "LastLoggedUser";
+auto constexpr kSettingsRating = "UserEditorRating";
+auto constexpr kSettingsChangesCount = "UserEditorChangsCount";
```
Comment (yunikkk) [inline editor/user_stats.cpp]: s/UserEditorChangsCount/UserEditorChangesCount

File: editor/editor_tests/user_stats_test.cpp (hunk 2 above)
Comment (mgsergio) [inline editor/editor_tests/user_stats_test.cpp]: Comment was added.

File: editor/user_stats.cpp (hunk 3 above)
Comment (mgsergio) [inline editor/user_stats.cpp]: Done (LastLoggedUser).

File: editor/user_stats.cpp (hunk 4 above)
Comment (mgsergio) [inline editor/user_stats.cpp]: Fixed: user stat info.

File: editor/user_stats.cpp (hunk 5 above)
Comment (mgsergio) [inline editor/user_stats.cpp]: Fixed: user stat info.

File: editor/user_stats.hpp [hunk 11]
```diff
@@ -1,30 +1,66 @@
...
+  void DropSettings();
+
+  string m_userName;
+
+  time_t m_lastUpdate{};
+  mutable mutex m_mutex;
 
-  /// True if last update was successful.
-  bool m_updateStatus;
+  shared_ptr<UserStats> m_userStats;
```
Comment (ygorshenin) [inline editor/user_stats.hpp]: Why do you need shared_ptr here? Just keep it by value, and return it by value where it's needed.

File: editor/user_stats.cpp [hunk 12]
```diff
@@ -60,9 +89,100 @@ bool UserStats::Update()
...
+  }).detach();
+}
+
+void UserStatsLoader::DropStats(string const & userName)
+{
+  lock_guard<mutex> g(m_mutex);
+  if (m_userName != userName)
+    return;
+  m_userStats = nullptr;
+  DropSettings();
```
Comment (ygorshenin) [inline editor/user_stats.cpp]: Do we need to call DropSettings with mutex?

File: editor/user_stats.cpp (hunk 6 above)
Comment (mgsergio) [inline editor/user_stats.cpp]: Fixed.

File: editor/user_stats.cpp (hunk 7 above)
Comment (mgsergio) [inline editor/user_stats.cpp]: Fixed.

File: editor/user_stats.cpp (hunk 8 above)
Comment (mgsergio) [inline editor/user_stats.cpp]: Detach in future destructor is guarantied only for c++14 
http://en.cppreference.com/w/cpp/thread/future/%7Efuture

File: map/framework.hpp (hunk 9 above)
Comment (mgsergio) [inline map/framework.hpp]: Fixed.
There is no need to call it. Fn() — is a signal to read a new data.

File: editor/user_stats.cpp (hunk 10 above)
Comment (mgsergio) [inline editor/user_stats.cpp]: Done.

File: editor/user_stats.hpp (hunk 11 above)
Comment (mgsergio) [inline editor/user_stats.hpp]: Fixed.

**Comment by mgsergio** [comment]
PTAL

**Comment by mgsergio** [comment]
DUDE JTTP

**Comment by mapsmetest** [comment]
Build started sha1 is merged.

File: editor/user_stats.cpp [hunk 13]
```diff
@@ -60,9 +99,101 @@ bool UserStats::Update()
...
+
+  if (nothingToUpdate)
+  {
+    GetPlatform().RunOnGuiThread(fn);
+    return;
+  }
+
+  thread([this, userName, &fn] {
+    if (Update(userName))
+      GetPlatform().RunOnGuiThread(fn);
```
Comment (igrechuhin) [inline editor/user_stats.cpp]: Если передать в качестве fn лямбду, она доживет до вызова?

File: editor/user_stats.hpp [hunk 14]
```diff
@@ -1,30 +1,69 @@
...
+  using TOnUpdateCallback = function<void()>;
+
+  /// Synchronously sends request to the server. Updates stats and returns true on success.
+  bool Update(string const & userName);
+  /// Launch the update process if stats are too old.
+  /// The process posts fn to a gui thread on success.
+  void Update(string const & userName, TOnUpdateCallback const & fn);
+
+  /// Resets internal state and removes records from settings.
+  void DropStats(string const & userNamae);
```
Comment (igrechuhin) [inline editor/user_stats.hpp]: userName

Comment (mgsergio) [inline editor/user_stats.hpp]: Done.

File: editor/user_stats.cpp (hunk 13 above)
Comment (mgsergio) [inline editor/user_stats.cpp]: Неа), я починил, спасибо. что заметил.

**Comment by mgsergio** [comment]
DUDE JTTP

File: editor/user_stats.hpp [hunk 15]
```diff
@@ -1,30 +1,69 @@
...
+  UserStats();
+  UserStats(time_t const updateTime, uint32_t const rating, uint32_t const changesCount);
 
-  bool IsChangesCountInitialized() const;
-  bool IsRankInitialized() const;
+  bool IsValid() const { return m_valid; }
 
-  int32_t GetChangesCount() const { return m_changesCount; }
-  int32_t GetRank() const { return m_rank; }
+  operator bool() const { return IsValid(); }
```
Comment (ygorshenin) [inline editor/user_stats.hpp]: Please, remove this.

**Comment by mapsmetest** [comment]
GitHub Watchdog Failure https://github.com/mapsme/omim/pull/3387

**Comment by mapsmetest** [comment]
Build started sha1 is merged.

File: editor/user_stats.hpp [hunk 16]
```diff
@@ -1,30 +1,69 @@
...
+class UserStatsLoader
+{
+public:
+  UserStatsLoader();
+
+  using TOnUpdateCallback = function<void()>;
+
+  /// Synchronously sends request to the server. Updates stats and returns true on success.
+  bool Update(string const & userName);
+  /// Launch the update process if stats are too old.
```
Comment (ygorshenin) [inline editor/user_stats.hpp]: A blank line, please.

File: editor/user_stats.hpp [hunk 17]
```diff
@@ -1,30 +1,69 @@
...
+  time_t m_updateTime;
+  bool m_valid;
+};
+
+class UserStatsLoader
+{
+public:
+  UserStatsLoader();
+
+  using TOnUpdateCallback = function<void()>;
```
Comment (ygorshenin) [inline editor/user_stats.hpp]: typedef should go first.

File: editor/user_stats.hpp [hunk 18]
```diff
@@ -1,30 +1,69 @@
...
+private:
+  /// Not thread-safe, but called only in constructor.
+  bool LoadFromSettings();
+  /// Not thread-safe, use synchonization.
+  void SaveToSettings();
+  void DropSettings();
+
+  string m_userName;
+
+  time_t m_lastUpdate{};
```
Comment (ygorshenin) [inline editor/user_stats.hpp]: Looks like this field is duplicated in m_userStats. Is it ok to remove it?

File: platform/platform.hpp [hunk 19]
```diff
@@ -126,6 +126,8 @@ class Platform
   /// @return path for directory in the persistent memory, can be the same
   /// as WritableDir, but on some platforms it's different
   string SettingsDir() const { return m_settingsDir; }
+  /// Set settings dir — use for testing.
+  void SetSettingsDirForTesting(string const & path);
```
Comment (ygorshenin) [inline platform/platform.hpp]: s/Testing/Tests/g

File: platform/platform_tests_support/writable_dir_changer.hpp [hunk 20]
```diff
@@ -1,12 +1,19 @@
 #include "std/string.hpp"
 
+enum class SettingsDirPolicy
```
Comment (ygorshenin) [inline platform/platform_tests_support/writable_dir_changer.hpp]: Move to the WritableDirChanger, please.

**Comment by mapsmetest** [comment]
GitHub Watchdog Failure https://github.com/mapsme/omim/pull/3387

File: editor/editor_tests/user_stats_test.cpp [hunk 21]
```diff
@@ -2,19 +2,56 @@
...
 
+#include "platform/platform_tests_support/writable_dir_changer.hpp"
+
 namespace editor
 {
 namespace
 {
-UNIT_TEST(UserStats_Smoke)
+auto constexpr kEditorTestDir = "editor-tests";
+WritableDirChanger wdc(kEditorTestDir, SettingsDirPolicy::UseWritableDir);
```
Comment (ygorshenin) [inline editor/editor_tests/user_stats_test.cpp]:
```c++
s/wdc/g_wdc/g

Comment (ygorshenin) [inline editor/editor_tests/user_stats_test.cpp]: Do you really want to change dir for all tests in editor_tests? I guess it's a bad idea.

Comment (ygorshenin) [inline editor/editor_tests/user_stats_test.cpp]: Because I would be a headache in other editor tests modules to realize why writable dir is changed.

File: editor/user_stats.cpp [hunk 22]
```diff
@@ -1,42 +1,81 @@
...
 string const kUserStatsUrl = "http://py.osmz.ru/mmwatch/user?format=xml";
-auto constexpr kUninitialized = -1;
+int32_t constexpr kUninitialized = -1;
+
+auto constexpr kSettingsUserName = "LastLoggedUser";
+auto constexpr kSettingsRating = "UserEditorRating";
+auto constexpr kSettingsChangesCount = "UserEditorChangesCount";
+auto constexpr kSettingsLastUpdate = "UserEditorLastUpdate";
+
+auto constexpr kOneHour = 60 * 60;
```
Comment (ygorshenin) [inline editor/user_stats.cpp]: s/kOneHour/kSecondsInHour/g

File: editor/user_stats.cpp [hunk 23]
```diff
@@ -60,9 +99,101 @@ bool UserStats::Update()
     return false;
   }
 
-  m_changesCount = document.select_node("mmwatch/edits/@value").attribute().as_int(kUninitialized);
-  m_rank = document.select_node("mmwatch/rank/@value").attribute().as_int(kUninitialized);
+  auto changesCount = document.select_node("mmwatch/edits/@value").attribute().as_int(-1);
+  auto rank = document.select_node("mmwatch/rank/@value").attribute().as_int(-1);
+
+  lock_guard<mutex> g(m_mutex);
+  if (m_userName != userName)
```
Comment (ygorshenin) [inline editor/user_stats.cpp]: Beware the ABA problem.

File: editor/user_stats.hpp (hunk 16 above)
Comment (mgsergio) [inline editor/user_stats.hpp]: Done.

File: editor/user_stats.hpp (hunk 17 above)
Comment (mgsergio) [inline editor/user_stats.hpp]: Ok

File: map/framework.cpp [hunk 24]
```diff
@@ -410,6 +410,8 @@ Framework::Framework()
   editor.LoadMapEdits();
 
   m_model.GetIndex().AddObserver(editor);
+
+  LOG(LINFO, ("Editor initialized"));
```
Comment (ygorshenin) [inline map/framework.cpp]: Do we really need this message?

File: platform/platform.hpp (hunk 19 above)
Comment (mgsergio) [inline platform/platform.hpp]: Done.

File: platform/platform_tests_support/writable_dir_changer.hpp (hunk 20 above)
Comment (mgsergio) [inline platform/platform_tests_support/writable_dir_changer.hpp]: Done.

File: platform/platform.cpp [hunk 25]
```diff
@@ -87,6 +87,11 @@ bool Platform::RmDirRecursively(string const & dirName)
   return res;
 }
 
+void Platform::SetSettingsDirForTesting(string const & path)
+{
+  m_settingsDir = my::AddSlashIfNeeded(path);
```
Comment (ygorshenin) [inline platform/platform.cpp]: Why do we need this? All methods that use settings dir must use JoinFoldersToPath, which joins path components with a slash.

Comment (mgsergio) [inline platform/platform.cpp]: SetWritebleDirForTests was implemented this way, I've just made it alike.

File: editor/user_stats.cpp (hunk 22 above)
Comment (mgsergio) [inline editor/user_stats.cpp]: Done.

File: editor/editor_tests/user_stats_test.cpp (hunk 21 above)
Comment (mgsergio) [inline editor/editor_tests/user_stats_test.cpp]: +1, thanks for saving from a headache!

**Comment by ygorshenin** [comment]
LGTM

**Comment by igrechuhin** [comment]
LGTM

**Comment by yunikkk** [comment]
LGTM

File: editor/user_stats.hpp [hunk 26]
```diff
@@ -1,30 +1,74 @@
...
 
 namespace editor
 {
 class UserStats
 {
 public:
-  explicit UserStats(string const & userName);
+  UserStats();
+  UserStats(time_t const updateTime, uint32_t const rating,
+            uint32_t const changesCount, string const & lelevUpFeat);
```
Comment (yunikkk) [inline editor/user_stats.hpp]: s/lelevUpFeat/levelUpFeat

Comment (mgsergio) [inline editor/user_stats.hpp]: Fixed.

**Comment by yunikkk** [comment]
LGTM again

**Comment by igrechuhin** [comment]
LGTM
